src/server/main_service.cc

```diff
@@ -68,6 +68,118 @@ std::optional<VarzFunction> engine_varz;
 
 constexpr size_t kMaxThreadSize = 1024;
 
+void UnregisterMonitors(ConnectionContext* server_ctx) {
+  ConnectionState& conn_state = server_ctx->conn_state;
+  if (conn_state.monitor) {
+    // remove monitor on this connection
+    auto token = conn_state.monitor->borrow_token;
+    server_ctx->ChangeMonitor(false /*start*/);
+    token.Wait();
+  }
+}
+
+// The format of the message that are sending is
+// +"time of day" [db-number <lua|unix:path|connection info] "command" "arg1" .. "argM"
+std::string MonitorTimestamp() {
+  timeval tv;
+
+  gettimeofday(&tv, nullptr);
+  return absl::StrCat(tv.tv_sec, ".", tv.tv_usec, absl::kZeroPad6);
+}
+
+auto CmdEntryToMonitorFormat(std::string_view str) -> std::string {
+  // This code is based on Redis impl for it at sdscatrepr@sds.c
+  std::string result = absl::StrCat("\"");
+
+  for (auto c : str) {
+    switch (c) {
+      case '\\':
+        absl::StrAppend(&result, "\\\\");
+        break;
+      case '"':
+        absl::StrAppend(&result, "\\\"");
+        break;
+      case '\n':
+        absl::StrAppend(&result, "\\n");
+        break;
+      case '\r':
+        absl::StrAppend(&result, "\\r");
+        break;
+      case '\t':
+        absl::StrAppend(&result, "\\t");
+        break;
+      case '\a':
+        absl::StrAppend(&result, "\\a");
+        break;
+      case '\b':
+        absl::StrAppend(&result, "\\b");
+        break;
+      default:
+        if (isprint(c)) {
+          result += c;
+        } else {
+          absl::StrAppend(&result, "\\x", absl::Hex((unsigned char)c, absl::kZeroPad2));
+        }
+        break;
+    }
+  }
+  absl::StrAppend(&result, "\"");
+  return result;
+}
+
+std::string MakeMonitorMessage(const ConnectionState& conn_state,
+                               const facade::Connection* connection, CmdArgList args) {
+  std::string message = MonitorTimestamp();
+
+  auto AppendConnDetails = [&]() {
+    if (conn_state.script_info.has_value()) {
+      absl::StrAppend(&message, "lua] ");
+    } else {
+      absl::StrAppend(&message, connection->RemoteEndpointStr());
+    }
+  };
+  auto AppendCmd = [&]() {
+    if (args.empty()) {
+      absl::StrAppend(&message, "error - empty cmd list!");
+    } else if (auto cmd_name = std::string_view(args[0].data(), args[0].size());
+               cmd_name == "AUTH") {  // we cannot just send auth details in this case
+      absl::StrAppend(&message, "\"", cmd_name, "\"");
+    } else {
+      message = std::accumulate(args.begin(), args.end(), message, [](auto str, const auto& cmd) {
+        absl::StrAppend(&str, " ",
+                        CmdEntryToMonitorFormat(std::string_view(cmd.data(), cmd.size())));
+        return str;
+      });
+    }
+  };
+
+  AppendConnDetails();
+  AppendCmd();
+  return message;
+}
+
+void MonitorCmd(bool admin_cmd, ConnectionContext* connection, CmdArgList args) {
```

Comment:
I find naming confusing. it's not a function handling `Monitor` command. Maybe `DispatchMonitorIfNeeded` ?

Reply:
OK - naming is subjective, if you think that this would make it more clean I will rename the function